diff -r a6e10414311a libgo/Makefile.am
--- a/libgo/Makefile.am	Fri Jan 16 07:57:02 2015 -0800
+++ b/libgo/Makefile.am	Fri Apr 03 15:07:47 2015 -0700
@@ -3738,7 +3738,6 @@
 	os/check \
 	path/check \
 	reflect/check \
-	regexp/check \
 	runtime/check \
 	sort/check \
 	strconv/check \
@@ -3838,7 +3837,6 @@
 	os/user/check \
 	path/filepath/check \
 	regexp/syntax/check \
-	runtime/pprof/check \
 	sync/atomic/check \
 	text/scanner/check \
 	text/tabwriter/check \
diff -r a6e10414311a libgo/Makefile.in
--- a/libgo/Makefile.in	Fri Jan 16 07:57:02 2015 -0800
+++ b/libgo/Makefile.in	Fri Apr 03 15:07:47 2015 -0700
@@ -2212,7 +2212,6 @@
 	os/check \
 	path/check \
 	reflect/check \
-	regexp/check \
 	runtime/check \
 	sort/check \
 	strconv/check \
@@ -2312,7 +2311,6 @@
 	os/user/check \
 	path/filepath/check \
 	regexp/syntax/check \
-	runtime/pprof/check \
 	sync/atomic/check \
 	text/scanner/check \
 	text/tabwriter/check \
diff -r a6e10414311a libgo/go/runtime/mfinal_test.go
--- a/libgo/go/runtime/mfinal_test.go	Fri Jan 16 07:57:02 2015 -0800
+++ b/libgo/go/runtime/mfinal_test.go	Fri Apr 03 15:07:47 2015 -0700
@@ -62,6 +62,7 @@
 		}()
 		<-done
 		runtime.GC()
+		runtime.GC()
 		select {
 		case <-ch:
 		case <-time.After(time.Second * 4):
