From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Chandler Carruth <chandlerc@gmail.com>
Date: Mon, 11 Oct 2021 02:06:57 +0000
Subject: Force every package to default to testonly.


diff --git a/absl/BUILD.bazel b/absl/BUILD.bazel
index 7239b5b..65de202 100644
--- a/absl/BUILD.bazel
+++ b/absl/BUILD.bazel
@@ -13,7 +13,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/algorithm/BUILD.bazel b/absl/algorithm/BUILD.bazel
index a3002b7..6c5d75a 100644
--- a/absl/algorithm/BUILD.bazel
+++ b/absl/algorithm/BUILD.bazel
@@ -22,7 +22,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/base/BUILD.bazel b/absl/base/BUILD.bazel
index 8b09b6e..98740a6 100644
--- a/absl/base/BUILD.bazel
+++ b/absl/base/BUILD.bazel
@@ -22,7 +22,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/cleanup/BUILD.bazel b/absl/cleanup/BUILD.bazel
index 5cca898..be4d182 100644
--- a/absl/cleanup/BUILD.bazel
+++ b/absl/cleanup/BUILD.bazel
@@ -20,7 +20,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/container/BUILD.bazel b/absl/container/BUILD.bazel
index 9e3bc06..bb13302 100644
--- a/absl/container/BUILD.bazel
+++ b/absl/container/BUILD.bazel
@@ -22,7 +22,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/debugging/BUILD.bazel b/absl/debugging/BUILD.bazel
index b503da5..3eaf1a1 100644
--- a/absl/debugging/BUILD.bazel
+++ b/absl/debugging/BUILD.bazel
@@ -23,6 +23,7 @@ load(
 )
 
 package(
+    default_testonly = 1,
     default_visibility = ["//visibility:public"],
 )
 
diff --git a/absl/flags/BUILD.bazel b/absl/flags/BUILD.bazel
index 7957c6d..8eca9b4 100644
--- a/absl/flags/BUILD.bazel
+++ b/absl/flags/BUILD.bazel
@@ -22,7 +22,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/functional/BUILD.bazel b/absl/functional/BUILD.bazel
index 4555cd5..d3aa1c9 100644
--- a/absl/functional/BUILD.bazel
+++ b/absl/functional/BUILD.bazel
@@ -22,7 +22,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/hash/BUILD.bazel b/absl/hash/BUILD.bazel
index 392b68f..f015011 100644
--- a/absl/hash/BUILD.bazel
+++ b/absl/hash/BUILD.bazel
@@ -22,7 +22,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/memory/BUILD.bazel b/absl/memory/BUILD.bazel
index d2824a0..898e756 100644
--- a/absl/memory/BUILD.bazel
+++ b/absl/memory/BUILD.bazel
@@ -22,7 +22,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/meta/BUILD.bazel b/absl/meta/BUILD.bazel
index 5585fcc..526287e 100644
--- a/absl/meta/BUILD.bazel
+++ b/absl/meta/BUILD.bazel
@@ -22,7 +22,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/numeric/BUILD.bazel b/absl/numeric/BUILD.bazel
index ea587bf..3cec06d 100644
--- a/absl/numeric/BUILD.bazel
+++ b/absl/numeric/BUILD.bazel
@@ -20,7 +20,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/profiling/BUILD.bazel b/absl/profiling/BUILD.bazel
index 496a06b..548587b 100644
--- a/absl/profiling/BUILD.bazel
+++ b/absl/profiling/BUILD.bazel
@@ -19,7 +19,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:private"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:private"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/random/BUILD.bazel b/absl/random/BUILD.bazel
index 66ffcbc..1e9325e 100644
--- a/absl/random/BUILD.bazel
+++ b/absl/random/BUILD.bazel
@@ -24,7 +24,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/random/internal/BUILD.bazel b/absl/random/internal/BUILD.bazel
index df6e42f..bb24ae3 100644
--- a/absl/random/internal/BUILD.bazel
+++ b/absl/random/internal/BUILD.bazel
@@ -26,9 +26,12 @@ load(
     "absl_random_randen_copts_init",
 )
 
-package(default_visibility = [
-    "//absl/random:__pkg__",
-])
+package(
+    default_testonly = 1,
+    default_visibility = [
+        "//absl/random:__pkg__",
+    ],
+)
 
 licenses(["notice"])
 
diff --git a/absl/status/BUILD.bazel b/absl/status/BUILD.bazel
index 30df22a..d253ac2 100644
--- a/absl/status/BUILD.bazel
+++ b/absl/status/BUILD.bazel
@@ -24,7 +24,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/strings/BUILD.bazel b/absl/strings/BUILD.bazel
index c45a671..83e37a0 100644
--- a/absl/strings/BUILD.bazel
+++ b/absl/strings/BUILD.bazel
@@ -21,6 +21,7 @@ load(
 )
 
 package(
+    default_testonly = 1,
     default_visibility = ["//visibility:public"],
     features = ["parse_headers"],
 )
diff --git a/absl/synchronization/BUILD.bazel b/absl/synchronization/BUILD.bazel
index 92e2448..a523592 100644
--- a/absl/synchronization/BUILD.bazel
+++ b/absl/synchronization/BUILD.bazel
@@ -22,7 +22,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/time/BUILD.bazel b/absl/time/BUILD.bazel
index 3e25ca2..b50b5e5 100644
--- a/absl/time/BUILD.bazel
+++ b/absl/time/BUILD.bazel
@@ -22,7 +22,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/time/internal/cctz/BUILD.bazel b/absl/time/internal/cctz/BUILD.bazel
index f549af6..b79cec1 100644
--- a/absl/time/internal/cctz/BUILD.bazel
+++ b/absl/time/internal/cctz/BUILD.bazel
@@ -14,7 +14,10 @@
 
 load("@rules_cc//cc:defs.bzl", "cc_library", "cc_test")
 
-package(features = ["-parse_headers"])
+package(
+    default_testonly = 1,
+    features = ["-parse_headers"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/types/BUILD.bazel b/absl/types/BUILD.bazel
index 83be936..e75ed2e 100644
--- a/absl/types/BUILD.bazel
+++ b/absl/types/BUILD.bazel
@@ -21,7 +21,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
diff --git a/absl/utility/BUILD.bazel b/absl/utility/BUILD.bazel
index 02b2c40..699cb43 100644
--- a/absl/utility/BUILD.bazel
+++ b/absl/utility/BUILD.bazel
@@ -22,7 +22,10 @@ load(
     "ABSL_TEST_COPTS",
 )
 
-package(default_visibility = ["//visibility:public"])
+package(
+    default_testonly = 1,
+    default_visibility = ["//visibility:public"],
+)
 
 licenses(["notice"])
 
-- 
2.33.0.882.g93a45727a2-goog

