Index: lib/CodeGen/BackendUtil.cpp
===================================================================
--- lib/CodeGen/BackendUtil.cpp	(revision 191330)
+++ lib/CodeGen/BackendUtil.cpp	(working copy)
@@ -336,6 +336,9 @@
       MPM->add(createStripSymbolsPass(true));
   }
 
+  // Force MergeFunctions pass.
+  MPM->add(createMergeFunctionsPass());
+
   PMBuilder.populateModulePassManager(*MPM);
 }
 
