Index: lib/AST/DeclObjC.cpp
===================================================================
--- lib/AST/DeclObjC.cpp	(revision 152265)
+++ lib/AST/DeclObjC.cpp	(working copy)
@@ -330,6 +330,10 @@
     LoadExternalDefinition();
 
   while (ClassDecl != NULL) {
+    // FIXME: Should make sure no callers ever do this.
+    if (!ClassDecl->hasDefinition())
+      return 0;
+        
     if ((MethodDecl = ClassDecl->getMethod(Sel, isInstance)))
       return MethodDecl;
 
