Index: include/clang/Basic/ConvertUTF.h
===================================================================
--- include/clang/Basic/ConvertUTF.h	(revision 152265)
+++ include/clang/Basic/ConvertUTF.h	(working copy)
@@ -135,6 +135,7 @@
   const UTF8** sourceStart, const UTF8* sourceEnd,
   UTF32** targetStart, UTF32* targetEnd, ConversionFlags flags);
 
+#define CLANG_NEEDS_THESE_ONE_DAY // LLDB wants these functions
 #ifdef CLANG_NEEDS_THESE_ONE_DAY
 ConversionResult ConvertUTF16toUTF8 (
   const UTF16** sourceStart, const UTF16* sourceEnd,
