--- a/compiler-rt/BUILD.bazel
+++ b/compiler-rt/BUILD.bazel
@@ -0,0 +1,18 @@
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "Fuzzer",
+    srcs = glob(
+        ["lib/fuzzer/Fuzzer*.cpp"],
+    ),
+    hdrs = glob([
+        "lib/fuzzer/Fuzzer*.h",
+        "lib/fuzzer/Fuzzer*.def",
+    ]),
+    copts = [
+      # FuzzerTracePC.h:196:12: runtime error: applying non-zero offset 8 to null pointer
+      "-fno-sanitize=undefined",
+    ],
+    includes = ["lib/fuzzer"],
+)
+
