// RUN: %clang_cc1 -fsyntax-only -verify -x c -std=c11 %s
// RUN: %clang_cc1 -fsyntax-only -verify -x c++ -std=c++11 %s
// RUN: %clang_cc1 -E -DPP_ONLY=1 %s -o %t
// RUN: FileCheck --strict-whitespace --input-file=%t %s

// This file contains Unicode characters; please do not "fix" them!

extern intÂ x; // expected-warning {{treating Unicode character as whitespace}}
extern intã€€x; // expected-warning {{treating Unicode character as whitespace}}

// CHECK: extern intÂ {{x}}
// CHECK: extern intã€€{{x}}

#pragma mark Â¡Unicode!

#define COPYRIGHT Copyright Â© 2012
#define XSTR(X) #X
#define STR(X) XSTR(X)

static const char *copyright = STR(COPYRIGHT); // no-warning
// CHECK: static const char *copyright = "Copyright Â© {{2012}}";

#if PP_ONLY
COPYRIGHT
// CHECK: Copyright Â© {{2012}}
CHECK: The preprocessor should not complain about Unicode characters like Â©.
#endif

// A ðŸŒ¹ by any other name....
extern int ðŸŒ¹;
int ðŸŒµ(int ðŸŒ») { return ðŸŒ»+ 1; }
int main () {
  int ðŸŒ· = ðŸŒµ(ðŸŒ¹);
  return ðŸŒ·;
}
