// RUN: %clang_cc1 -fsyntax-only -verify -x c -std=c11 %s
// RUN: %clang_cc1 -fsyntax-only -verify -x c++ -std=c++11 %s
// RUN: %clang_cc1 -E -DPP_ONLY=1 %s -o %t
// RUN: FileCheck --strict-whitespace --input-file=%t %s

// This file contains Unicode characters; please do not "fix" them!

extern intÂ x; // expected-warning {{treating Unicode character as whitespace}}
extern intã€€x; // expected-warning {{treating Unicode character as whitespace}}

// CHECK: extern intÂ {{x}}
// CHECK: extern intã€€{{x}}

#pragma mark Â¡Unicode!

#define COPYRIGHT Copyright Â© 2012
#define XSTR(X) #X
#define STR(X) XSTR(X)

static const char *copyright = STR(COPYRIGHT); // no-warning
// CHECK: static const char *copyright = "Copyright Â© {{2012}}";

#if PP_ONLY
COPYRIGHT
// CHECK: Copyright Â© {{2012}}
CHECK: The preprocessor should not complain about Unicode characters like Â©.
#endif

// A ğŸŒ¹ by any other name....
extern int ğŸŒ¹;
int ğŸŒµ(int ğŸŒ») { return ğŸŒ»+ 1; }
int main () {
  int ğŸŒ· = ğŸŒµ(ğŸŒ¹);
  return ğŸŒ·;
}

int nÍ¾ = 3; // expected-warning {{treating Unicode character <U+037E> as identifier character rather than as ';' symbol}}
int *nê‰ê‰v = &nÍ¾; // expected-warning 2{{treating Unicode character <U+A789> as identifier character rather than as ':' symbol}}
                 // expected-warning@-1 {{treating Unicode character <U+037E> as identifier character rather than as ';' symbol}}
int vï¼ï¼»ï¼ï¼½ï¼ˆautoï¼‰ï½›returnï½xï¼›ï½ï¼ˆï¼‰; // expected-warning 12{{treating Unicode character}}

int â xï»¿xâ€;
// expected-warning@-1 {{identifier contains Unicode character <U+2060> that is invisible in some environments}}
// expected-warning@-2 {{identifier contains Unicode character <U+FEFF> that is invisible in some environments}}
// expected-warning@-3 {{identifier contains Unicode character <U+200D> that is invisible in some environments}}
int fooâ€‹bar = 0; // expected-warning {{identifier contains Unicode character <U+200B> that is invisible in some environments}}
int x = foobar; // expected-error {{undeclared identifier}}

int âˆ£foo; // expected-error {{non-ASCII character}}
#ifndef PP_ONLY
#define âˆ¶ x // expected-error {{macro name must be an identifier}}
#endif
